core/src/main/scala/org/ensime/core/DebugManager.scala

```diff
@@ -237,67 +239,75 @@ class DebugManager(
     case e: MethodExitEvent =>
   }
 
-  def handleStartupFailure(e: Exception): Unit = {
+  def disposeCurrentVM(): Unit = {
+    withVM { vm =>
+      vm.dispose()
+    }
+  }
+  def handleStartupFailure(e: Exception): RpcResponse = {
     maybeVM = None
     log.error(e, "Failure during VM startup")
     val message = e.toString
-    sender ! DebugVmError(1, message)
+    DebugVmError(1, message)
+  }
+
+  def handleDebugStartReq(commandLine: String): RpcResponse = {
+    disposeCurrentVM()
+    try {
+      val vm = new VM(VmStart(commandLine))
+      maybeVM = Some(vm)
+      vm.start()
+      DebugVmSuccess()
+    } catch {
+      case e: Exception =>
+        log.error(e, "Could not start VM")
+        handleStartupFailure(e)
+    }
+  }
+
+  def handleDebugAttachReq(hostname: String, port: String): RpcResponse = {
+    disposeCurrentVM()
+    try {
+      val vm = new VM(VmAttach(hostname, port))
+      maybeVM = Some(vm)
+      vm.start()
+      DebugVmSuccess()
+    } catch {
+      case e: Exception =>
+        log.error(e, "Could not attach VM")
+        handleStartupFailure(e)
+    }
   }
 
   def fromUser: Receive = {
-    case DebugStartReq(commandLine: String) ⇒
-      withVM { vm ⇒
-        vm.dispose()
-      }
-      try {
-        val vm = new VM(VmStart(commandLine))
-        maybeVM = Some(vm)
-        vm.start()
-        sender ! DebugVmSuccess()
-      } catch {
-        case e: Exception =>
-          log.error(e, "Could not start VM")
-          handleStartupFailure(e)
-      }
+    case DebugStartReq(commandLine: String) =>
+      sender ! handleDebugStartReq(commandLine)
     case DebugAttachReq(hostname, port) ⇒
-      withVM { vm ⇒
-        vm.dispose()
-      }
-      try {
-        val vm = new VM(VmAttach(hostname, port))
-        maybeVM = Some(vm)
-        vm.start()
-        sender ! DebugVmSuccess()
-      } catch {
-        case e: Exception =>
-          log.error(e, "Could not attach VM")
-          handleStartupFailure(e)
-      }
+      sender ! handleDebugAttachReq(hostname, port)
     case DebugActiveVmReq =>
-      handleRPCWithVM() { vm =>
-        sender ! TrueResponse
+      sender ! handleRPCWithVM() { vm =>
+        TrueResponse
       }
-
     case DebugStopReq =>
-      handleRPCWithVM() { vm =>
+      sender ! handleRPCWithVM() { vm =>
+        if (vm.mode.shouldExit) {
```

Comment:
@edani This fix means it should only trigger the exit on a VM started by Ensime - as per original discussion.